[Feat] Dynamic data parameters automatically refresh the relevant dynamic filters (#1051)
mckinsey/vizro: vizro-core/docs/pages/user-guides/data.md

### Filters

> When a [filter](filters.md) depends on dynamic data and no `selector` is explicitly defined in the `vm.Filter` model, the available selector values update either when the page refreshes or when a relevant [dynamic data parameter](#parametrize-data-loading) changes. This ensures the filter always reflects the latest data, which is called a _dynamic filter_.

Reviewer: ```suggestion
When a [filter](filters.md) depends on dynamic data and no `selector` is explicitly defined in the `vm.Filter` model, it is called a _dynamic filter_. A dynamic filter always reflects the latest data since the available selector values update either when the page refreshes or when a relevant [dynamic data parameter](#parametrize-data-loading) changes.
```

For example, let us extend the [parametrized dynamic data example](#parametrize-data-loading) above with a filter. When you run the code and refresh the page, the filter and Graph update automatically. The same happens when the dynamic data parameter [`Slider`][vizro.models.Slider] changes, ensuring the latest data is always reflected.

!!! example "Dynamic filters"
    === "app.py"
        ```py hl_lines="20"
        from vizro import Vizro
        import pandas as pd
        import vizro.plotly.express as px
        import vizro.models as vm

        from vizro.managers import data_manager

        def load_iris_data(number_of_points=10):
            iris = pd.read_csv("iris.csv")
            return iris.sample(number_of_points)

        data_manager["iris"] = load_iris_data

        page = vm.Page(
            title="Update the chart on page refresh or when the Parameter changes",
            components=[
                vm.Graph(id="graph", figure=px.box("iris", x="species", y="petal_width", color="species"))
            ],
            controls=[

>                 vm.Filter(column="species", selector=vm.RadioItems()),  # (1)!

Author: Comparing from the previous example ("Parametrized dynamic data example"), this is the only line that's added.

                vm.Parameter(
                    targets=["graph.data_frame.number_of_points"],
                    selector=vm.Slider(min=1, max=10, step=1, value=1),
                )
            ],
        )

        dashboard = vm.Dashboard(pages=[page])

        Vizro().build(dashboard).run()
        ```

>         1. This filter implicitly controls the dynamic data source `"iris"`, which supplies the `data_frame` to the targeted `vm.Graph`. On page refresh or when the dynamic data parameter changes, Vizro reloads this data, finds all the unique values in the `"species"` column and sets the categorical selector's `options` accordingly.

Reviewer: ```suggestion
        1. This filter implicitly controls the dynamic data source `"iris"`, which supplies the `data_frame` to the targeted `vm.Graph`. On page refresh or when the dynamic data parameter changes, Vizro reloads this data, finds all the unique values in the `"species"` column and sets the categorical selector's `options` accordingly. We set  `selector=vm.RadioItems()` just to make the `options` more obvious in the example app; if `selector` had not been specified then the filter would have defaulted to using `vm.Dropdown` but options would refresh exactly the same way.
```
